ni-base-system-image-tests: Add filesystem permissions ptest

mike-petersen-ni opened the pull request:
As part of [AB#2232369](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/2232369), we want to notice when filesystem permissions change. Writing this as a ptest improves maintainability over the alternative of creating a standalone test, and allows future improvements (e.g. for isolation) to be easily utilized.

This ptest is not added to the ptest packagegroup, as the installation of other packages will alter its results.

# Behavior of the new test
This test is similar to the dmesg diff test at `recipes-kernel/kernel-tests`.

This test collects a manifest using `find`, using its `-printf` expression to generate a table in tab-separated-value format. To compare the tables, they are first converted to dictionaries and compared in Python, rather than using `diff` as in the dmesg diff test.

## Determining the previous manifest
Unlike the dmesg diff test, the kernel version is irrelevant. Instead, the OS version (retrieved from `/etc/os-release`) is used to fulfill the purpose. Since multiple variants may have similar versions (e.g. hardknott and kirkstone are now both building for 23.3.0), the codename is also extracted and used to distinguish between versions.

The version of the previous manifest must:
 - have major/minor/patch components that are lexically less than the current version
 - have a phase of "final"

It will be necessary to manually run this test on the last release version to create such a manifest to compare against.

## Directories examined
As seen in `fs_permissions_diff.py`:

```py
search_dirs = ['/bin', '/boot', '/etc', '/lib', '/lib64', '/sbin', '/usr', '/var']
omit_dirs = ['/lib/modules', '/var/cache', '/var/run', '/var/tmp', '/var/volatile']
```

Let me know if these lists should change (e.g. exclude `/etc/natinst/share`, include `/home` or `/tmp`).

## Handling of symbolic links
The test does not resolve symbolic links as their targets, though it does record the link target.  
Changing a link to/from a non-link will be noted as a difference, as that is part of the mode.  
Targets of links are not compared.

## Handling of added/removed paths
All paths present in the previous manifest but missing in the current one (or vice versa) are printed in the log and fail the test.

# Miscellaneous changes
 - dmesg diff test:
   - Removed unused import
   - Modified strip_headers to avoid bad asymptotic behavior

# Testing
 - [x] Built with `bitbake ni-base-system-image-tests`
 - [x] Installed on a NILRT VM with `opkg install ni-base-system-image-tests-ptest`
 - Ran `ptest-runner ni-base-system-image-tests` in various scenarios
   - [x] Verified that the manifest was uploaded to the Mongo database
   - Reran after changing content in `os-release` to...
      - [x] a later version; verified that the previous run was diffed against
      - [x] a different codename; verified that the previous run was *not* diffed against
    - [x] After changing permissions (mode, user, or group) on `/bin`, the test fails
    - [x] After adding files (equivalent: running without a previous manifest), the test fails
    - [x] After removing files, the test fails

# Merging
I think it is appropriate to cherry-pick this change into nilrt/master/hardknott.

Commit:
ni-base-system-image-tests: Copy files from other recipes

 - recipes-ni/nirtcfg-tests/nirtcfg-tests.bb to
   ni-base-system-image-tests.bb
 - recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py
   to files/fs_permissions_diff.py
 - recipes-kernel/kernel-tests/kernel-tests-files/ptest-format.sh to
   files/ptest-format.sh
 - recipes-kernel/kernel-tests/kernel-tests-files/test_kernel_dmesg_diff.sh
   to files/test_fs_permissions_diff.sh

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

Commit:
ni-base-system-image-tests: Create fs permissions diff test

As with the dmesg diff test, we want to have comparisons between
previous known-good versions and new versions--in this case, comparing
the filesystem permissions against the last released version of NILRT.

The test traverses most of the system directories and creates a
manifest, which is handled similarly to the dmesg log in that test. Any
change to permissions, or added/removed paths, provoke the test to fail.

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

chaitu236 reviewed (approved):
The changes look good but I don't understand this comment

> A squash merge is likely simplest for maintainers. The merge commit in the early history connects the histories across the file copies, but this means a rebase would require manual intervention (to avoid accidentally removing the files copied from).

Why would a rebase require manual intervention?

mike-petersen-ni commented:
> Why would a rebase require manual intervention?

It's an artifact of how I preserved the history across the file copies with a merge commit. Rebasing doesn't handle merge commits too well; it'd be necessary to do an interactive rebase that looks like this at the beginning:

```
label base
pick 93c79f67 Move nirtcfg-tests.bb from recipes-ni/nirtcfg-tests
pick a2ff13e7 Move kernel_dmesg_diff.py from recipes-kernel/kernel-tests
pick 0b00982e Move ptest-format.sh from recipes-kernel/kernel-tests
pick 4acfe422 Move test_kernel_dmesg_diff.sh from recipes-kernel/kernel-tests
label side
reset base
merge side
break # to manually un-delete the files with git checkout HEAD^ -- <files ...>
pick 1a84f40c Use the correct files in the recipe and tests
```

You could attempt to use `git rebase origin/nilrt/master/kirkstone --onto origin/nilrt/master/hardknott`, but you'd get a conflict later on since the files had been renamed instead of copied.

---

If y'all want to preserve my commit history here, but don't want to pull in this messy stuff, I could fix this problem by squashing the copies. I imagine we'd probably also have me edit the commit titles to have that "location: " prefix that's common in this repo.

chaitu236 commented:
> If y'all want to preserve my commit history here, but don't want to pull in this messy stuff, I could fix this problem by squashing the copies. I imagine we'd probably also have me edit the commit titles to have that "location: " prefix that's common in this repo.

Ok, I'll defer to Alex on how we want to handle this.

chaitu236 commented on recipes-ni/ni-base-system-image-tests/files/fs_permissions_diff.py:
```diff
@@ -0,0 +1,215 @@
+fs_manifest_columns = ['path', 'mode', 'user', 'group', 'link_target']
+def get_fs_manifest():
+    search_dirs = ['/bin', '/boot', '/etc', '/lib', '/lib64', '/sbin', '/usr', '/var']
+    omit_dirs = ['/var/cache', '/var/run', '/var/tmp', '/var/volatile']
```
It just occurred to me that kernel modules directory `/lib/modules/<kernelversion>` will have a new name on every update. Will this cause false positives? Do we need to omit that directory? I'm not sure if there is any benefit to indexing that directory.

mike-petersen-ni replied:
We could omit `/lib/modules` if we wanted to. As mentioned in the PR text added and removed files (and by extension, renames like this) are only logged, and won't fail the test. Let me know of your preference.

chaitu236 replied:
If we keep it, all the files within that directory will be logged right? That'd be too much noise so I'd prefer we not keep it.

mike-petersen-ni replied:
Makes sense. I'll push that change.

jpautler commented on recipes-ni/ni-base-system-image-tests/files/fs_permissions_diff.py:
```diff
@@ -0,0 +1,215 @@
+
+    def __init__(self, server, user, password):
+        mongo_client = pymongo.MongoClient("mongodb://{user}:{pwd}@{server}".format(
+            user=args.user, pwd=args.password, server=args.server
```
Did you meant to use "args." for these values? Assuming not, it will be a little confusing that the format string placeholder names are identical to the constructor parameter names, although I think it probably works anyway.

mike-petersen-ni replied:
Copied from original; I'll fix both with next push.

jpautler commented on recipes-ni/ni-base-system-image-tests/files/fs_permissions_diff.py:
```diff
@@ -0,0 +1,215 @@
+        return False
+
+    logger.log('INFO: Empty diff')
+    return True
```
Do you want this function to return False if there are **any** differences (including paths added or removed)? Currently, it seems to only returns False if a path is different.

mike-petersen-ni replied:
Do we want the test to fail for added/removed paths? I don't know if it'll provide much value to get a test failure from that.

jpautler replied:
Good point. Maybe a good question for @amstewart?

mike-petersen-ni replied:
Just pushed change to consider adds and removes as failures.

jpautler commented on recipes-ni/ni-base-system-image-tests/ni-base-system-image-tests.bb:
```diff
@@ -0,0 +1,34 @@
+# We only want to build the -ptest package
+PACKAGES:remove = "${PN}-dev ${PN}-staticdev ${PN}-dbg"
+
+RDEPENDS:${PN}-ptest:append = " bash python3-pymongo"
```
I thought maybe there should be a dependency on ptest-runner here, but I don't see it in the recipes for other ptest packages. Looks like it is included as a dependency in the packagegroups that install our ptests. Since this test is not part of a packagegroup, you might need to include the ptest-runner package in the optional_opkg_packages event variable so that it is present on your system in order to run the test.

mike-petersen-ni replied:
Addressed in the PR for the event.

amstewart commented:
>

amstewart commented:
> > If y'all want to preserve my commit history here, but don't want to pull in this messy stuff, I could fix this problem by squashing the copies. I imagine we'd probably also have me edit the commit titles to have that "location: " prefix that's common in this repo.
> 
> Ok, I'll defer to Alex on how we want to handle this.

Yeah; I'd prefer to squash this PR down into a smaller number of commits. @mike-petersen-ni Could you take care of that.

Commit:
kernel-tests: Fix minor issues

 - Use given arguments rather than overall arguments object
 - Remove unused import
 - Improve performance of strip_headers:
   Accumulating strings with

       result = result + something

   results in worse than O(n^2) behavior, due to string immutability.
   Using

       result += something

   is better, since it conveys that the destination is eligible for
   replacement, but it is better still to use string join, which
   presumably is smarter about allocations and copies, and is more
   likely to receive optimization attention.

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

mike-petersen-ni commented:
# Patch v5(?)

Conversed with @chaitu236 out-of-band; rebased to squash commits down to just three: fixup, copy, new work.

chaitu236 commented:
Cherry picked to hardknott

mike-petersen-ni commented on recipes-ni/ni-base-system-image-tests/files/fs_permissions_diff.py:
```diff
@@ -0,0 +1,239 @@
+    return ''.join(line + '\n' for line in fs_manifest.splitlines() if not line.startswith('#'))
+
+def get_old_fs_manifest(db, logger):
+    os_version = OsVersion()
+    query = {
+        'os_version_codename': os_version.codename,
+        'os_version_major_minor': os_version.major_minor,
+        'os_version': {
+            '$lt': {
+                'major': os_version.major,
+                'minor': os_version.minor,
+                'patch': os_version.patch,
+                # $lt considers booleans to be greater than all numbers
+                'phase': False,
+                'build': False
+            }
+        },
+        'os_version.phase': OsVersion.phase_number['f']
+    }
+
+    if db.count_documents(query):
+        results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
+    else:
+        # If there are no results, there may not be a run from the current os version.
+        # Remove that requirement.
+        del query['os_version_major_minor']
+        logger.log('INFO: No prior fs permissions from this OS Version found. Using previous version.')
+
+        if db.count_documents(query):
+            results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
+        else:
+            # Keep this log in first line to help streak indexer group results. Hashes will be populated at end.
+            logger.first_log('INFO: fs_permissions_diff: {} against <empty> '.format(os_version.full))
+
+            logger.log('INFO: No suitable previous fs permissions found')
+            return ''
```
Conversation with @jpautler on improving output for readers of diff output.

 - Consider having a double-diff (analogue: git's three-way merge) against a basis (version < maj.min.pat.0) and recent (version < maj.min.pat.bld) to consistently report differences, but also point out new ones.
  - Attempt to match codename if possible, but if not possible, relax that constraint (and assume there's only one alternative, or equivalently. highest version found will be in the "previous" codename)
  - Try sorting by full version instead of date

```suggestion
   os_version = OsVersion()
    query = {
        'os_version_codename': os_version.codename,
        'os_version': {
            '$lt': {
                'major': os_version.major,
                'minor': os_version.minor,
                'patch': os_version.patch,
                'build': 0
            }
        }
    }

    if db.count_documents(query):
        # TODO sort change
        results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
    else:
        # If there are no results, there may not be a run from the current os codename.
        # Remove that requirement.
        del query['os_version_codename']
        logger.log('INFO: No prior fs permissions from this codename found. Relaxing that constraint.')

        if db.count_documents(query):
            # TODO sort change
            results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
        else:
            # Keep this log in first line to help streak indexer group results. Hashes will be populated at end.
            logger.first_log('INFO: fs_permissions_diff: {} against <empty> '.format(os_version.full))

            logger.log('INFO: No suitable previous fs permissions found')
            return ''
```

Also make sure that `os_version.phase` isn't uploaded to MongoDB. (Do this before enabling event so that $lt doesn't have undesirable behavior with key count stuff.)

mike-petersen-ni replied:
Possible printout representation: similar to git's 3-way merge with two columns of +/-/space
